src/util/key.rs

```diff
@@ -109,15 +109,19 @@ impl PublicKey {
         }
     }
 
-    /// Returns bitcoin 160-bit hash of the public key
-    pub fn pubkey_hash(&self) -> PubkeyHash {
+    fn with_serialized<R, F: FnOnce(&[u8]) -> R>(&self, f: F) -> R {
```

Comment:
I'm not familial with this pattern but the `with_foo` naming made me think of the builder pattern. Perhaps this reads better?
```rust
    // Serializes `self` based on compressed'ness and calls `f`.
    fn serialize_and_call<R, F: FnOnce(&[u8]) -> R>(&self, f: F) -> R {...}

    self.serialize_and_call(|bytes| ...);
```